hardware/VirtualThermostat.cpp

```diff
@@ -0,0 +1,715 @@
+#include "stdafx.h"
+#include "../main/Helper.h"
+#include "../main/Logger.h"
+#include "../main/SQLHelper.h"
+#include "../main/RFXtrx.h"
+
+#include "VirtualThermostat.h"
+
+#include "../main/localtime_r.h"
+#include "../main/mainworker.h"
+#include "../notifications/NotificationHelper.h"
+#include <stdarg.h>     
+
+std::string GetDeviceValue(const char* FieldName, const char* Idx)
+{
+	auto result = m_sql.safe_query("SELECT %s from DeviceStatus WHERE (ID == %s )", FieldName, Idx);
+	if (!result.empty())
+		return  result[0][0];
+	else
+		return  "";
+}
+
+void VirtualThermostat::CircularBuffer::Clear()
+{
+	for (int i = 0; i < Size; i++)Value[i] = 0;
+	index = 0;
+	Sum = 0;
+}
+VirtualThermostat::CircularBuffer::CircularBuffer(int pSize)
+{
+	Value = new double[pSize];
```

Comment:
Do not use raw memory https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rr-newdelete
Use containers like `std::vector` instead.
But at the end, why not using something like https://www.boost.org/doc/libs/1_81_0/doc/html/circular_buffer.html instead of reinventing it?

Reply:
ok. but i dont know boost library !

Reply:
the allocation is done one time during reset and is not freed after
so it is more effecient than the std::vector that the allocation  is done during runtime

Reply:
I guess you would create a vector only once with all its elements initialized to 0 with
```c++
std::vector<double> Value(pSize);
```

Reply:
ok ! use circular buffer